core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1829,7 +1901,58 @@ private[spark] class BlockManager(
     data.dispose()
   }
 
+  /**
+   * Class to handle block manager decommissioning retries
+   * It creates a Thread to retry offloading all RDD cache blocks
+   */
+  private class BlockManagerDecommissionManager(conf: SparkConf) {
+    @volatile private var stopped = false
+    private val sleepInterval = conf.get(
+      config.STORAGE_DECOMMISSION_REPLICATION_REATTEMPT_INTERVAL)
+
+    private val blockReplicationThread = new Thread {
+      override def run(): Unit = {
+        var failures = 0
+        while (blockManagerDecommissioning
+          && !stopped
+          && !Thread.interrupted()
+          && failures < 20) {
+          try {
+            logDebug("Attempting to replicate all cached RDD blocks")
+            decommissionRddCacheBlocks()
+            logInfo("Attempt to replicate all cached blocks done")
+            Thread.sleep(sleepInterval)
+          } catch {
+            case _: InterruptedException =>
+              logInfo("Interrupted during migration, will not refresh migrations.")
+              stopped = true
+            case NonFatal(e) =>
+              failures += 1
+              logError("Error occurred while trying to replicate cached RDD blocks" +
+                s" for block manager decommissioning (failure count: $failures)", e)
+          }
+        }
+      }
+    }
+    blockReplicationThread.setDaemon(true)
+    blockReplicationThread.setName("block-replication-thread")
```

Comment:
Use `Runnable` for the decommissioning and `ThreadUtils` to execute the `Runnable`?

Reply:
Looking at our code we seem to be roughly split on `Thread` versus `Runnable` usage. I think `Runnable` would make more sense if we were submitting this to an execution pool, but since we have a single thread and there is no reason to scale up the number of threads I don't see the need for that change.

Reply:
We always use `ThreadUtils.newDaemonSingleThreadExecutor` for the single runnable.

Reply:
`grep -r "new Thread" ./core/src/main | wc -l` returns 36
`grep -r ThreadUtils.newDaemonSingleThreadExecutor ./core/src/main |wc -l` returns 4

Reply:
Ah..that's good point but just wondering how many of them are chosen after realizing `ThreadUtils.newDaemonSingleThreadExecutor`. 

BTW, you'd better grep "new Thread(" to exclude ThreadLocal declaration.

Reply:
Even that returns 36 in core.